[Weather][MeteoBlue] MeteoBlue weather binding

9037568 opened the pull request:
This is very early code for a new ESH implementation MeteoBlue weather binding. (For #4200).

Basic functionality (pulling data from the weather service and populating channels/items) is complete.

Things that still need to be looked into:
- images: need a way to build the weather icons into the binding
- conversions
- i18n
- port the Commercial API functionality over from OH1?

All feedback is most welcome.

9037568 commented:
I don't understand this travis build failure. It builds fine locally.

sjka commented:
It fails because you are referring to `0.9.0-SNAPSHOT` parent. However, Eclipse SmartHome is at `0.10.0-SNAPSHOT` currently. You still have those in your local maven repository. Travis is set up to be as clean as possible, therefore it doesn't.

triller-telekom commented:
> Things that still need to be looked into:
> * conversions

You can put this point at the back of your list until we have merged PR #4818. This PR will introduce a new DataType with which you can easily convert from one unit to another.

9037568 commented:
@sjka:

> It fails because you are referring to 0.9.0-SNAPSHOT parent. However, Eclipse SmartHome is at 0.10.0-SNAPSHOT currently

Thanks for this. But I guess I don't understand how to update this. If I change the versions in pom.xml and MANIFEST.MF, I get this error:

````
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact org.eclipse.smarthome:targetplatform:target:smarthome:0.10.0-SNAPSHOT
````

sjka commented:
I assume you are trying to build only your binding, right? This (of course) is a perfectly valid use-case normally, but as this project doesn't publish any snapshots on maven central you will have to build the target platform project once locally after such a version bump in order to get it into your local repository cache. See also https://github.com/eclipse/smarthome/issues/4140#issuecomment-334958255.

9037568 commented:
I think this is as good as it gets without some feedback.
I therefore call this an alpha version and request said feedback.

One thing I'm not particularly happy with is the various type definitions in the thing-types file. I'd like to not have to define 10000 temperature types, but simply define 1 and have it referenced by all the things that are of that type. However I've been unable to find a way to implement that.

kaikreuzer commented:
>  I'd like to not have to define 10000 temperature types, but simply define 1 and have it referenced by all the things that are of that type.

Not sure what you mean - from what I see, you only define each type exactly once. You simply have many different types, which all have a different semantic, so what is there to do?

lolodomo commented:
Does MeteoBlue provide free API key ?

9037568 commented:
@kaikreuzer I think it's a terminology problem. I don't call "min temperature" and "max temperature" two types; they're each an instance of a single "temperature" type.

It sounds like ESH doesn't have a way of defining types in a non-redundant way.

9037568 commented:
@lolodomo 
>Does MeteoBlue provide free API key ?

Yes, there's a non-commercial license for the basic-day API package.

kaikreuzer commented:
> I think it's a terminology problem.

Well, terminology is [defined here](https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels):

_The granularity of channel types should be on its semantic level, i.e. very fine-grained: If a Thing measures two temperature values, one for indoor and one for outdoor, this should be modelled as two different channel types._

> they're each an instance of a single "temperature" type.

No, they each link to `Number:Temperature` item type, but their meaning is something different.

9037568 commented:
>No, they each link to Number:Temperature item type, but their meaning is something different.

In that case I think there are still a few I need to change.

triller-telekom commented on extensions/binding/pom.xml:
```diff
@@ -38,6 +38,8 @@
     <module>org.eclipse.smarthome.binding.wemo</module>
     <module>org.eclipse.smarthome.binding.wemo.test</module>
     <module>org.eclipse.smarthome.binding.yahooweather</module>
+    <module>org.eclipse.smarthome.binding.meteoblue</module>
```
Please also add your binding to the karaf feature file: https://github.com/eclipse/smarthome/blob/master/features/karaf/esh-ext/src/main/feature/feature.xml

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,578 @@
+			<parameter name="refresh" type="integer" min="5" required="false" unit="min">
+				<label>Refresh interval</label>
+				<description>Refresh interval (in minutes)</description>
+				<default>240</default>
```
If I've understood [their documentation](https://content.meteoblue.com/en/research-development/data-sources/weather-modelling/model-run) correctly, the models are only updated twice a day, so I thought a default of 4 here would be more than adequate.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue.test/pom.xml:
```diff
@@ -0,0 +1,127 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.binding.meteoblue.test</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.binding.meteoblue</bundle.namespace>
+    <checkstyle.skip>true</checkstyle.skip>
```
Don't recall at the moment. Will recheck.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,578 @@
+	</channel-type>
+
+	<channel-type id="minTemperature">
+		<item-type>Number</item-type>
```
Where does one get the Channel info (i.e. the item-type, category, etc)? A rational guess would be to start by getting a Channel object and then calling `channel.getType()` but NO, that method isn't available.

9037568 commented:
Is this not a valid state definition? I thought this `pattern` should make the number show up in the UI as "97 %", "99 %", etc., but the " %" isn't shown.

````
<state readOnly="true" min="0" max="100" pattern="%d %%" />
````

9037568 commented:
That should do it for UoM, I think. Bridge work next...

9037568 commented:
I think this completes the bridge. Next I'll try to add examples to the README.

triller-telekom commented:
@9037568 Please let me know once you are done. And please also rebase you branch onto the current master to resolve the conflict in the `feature.xml`.

9037568 commented:
@triller-telekom Assuming Travis is happy, I'm done.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/internal/MeteoBlueHandlerFactory.java:
```diff
@@ -34,7 +38,8 @@
 public class MeteoBlueHandlerFactory extends BaseThingHandlerFactory {
 
-    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_WEATHER);
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Sets.union(
```
Should be possible to implement without an external library:
```
Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Stream.concat(
                    BRIDGE_THING_TYPES_UIDS.stream(),
                    SUPPORTED_THING_TYPES_UIDS.stream()).collect(toSet());
```

triller-telekom replied:
Would be nice if you remove the external dependency here.

9037568 replied:
It'd be nice if your suggested code compiled, but it doesn't.
I tried changing it to this, but it also doesn't compile:

````
Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS =
        Stream.concat(
        BRIDGE_THING_TYPES_UIDS.stream(),
        MeteoBlueBindingConstants.SUPPORTED_THING_TYPES_UIDS.stream()).collect(toSet());
````

````
[ERROR] MeteoBlueBindingConstants.SUPPORTED_THING_TYPES_UIDS.stream()).collect(toSet());
[ERROR] ^^^^^
[ERROR] The method toSet() is undefined for the type MeteoBlueHandlerFactory
````

triller-telekom replied:
Upps sorry, I missed the `Collectors.` in front of `toSet()`.

9037568 replied:
Yes, that appears to work.
Done.

triller-telekom commented on features/karaf/esh-ext/src/main/feature/feature.xml:
```diff
@@ -83,6 +83,11 @@
+    <feature>esh-base</feature>
+    <bundle>mvn:org.eclipse.smarthome.binding/org.eclipse.smarthome.binding.meteoblue/${project.version}</bundle>
+  </feature>
+
```
Please remove this blank line.

9037568 replied:
This one I don't get. There's a blank line after every `<feature>` block in the file. Why should this one be different?

triller-telekom replied:
Upps sorry, github only showed me a small portion of the file. You are right, sorry for the confusion.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -6,19 +6,20 @@ to provide weather information.
+its child things.
 
 
 ## Thing Configuration
```
Sure. Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -84,7 +89,7 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         // updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -228,10 +267,58 @@ private void updateChannel(String channelId) {
     // Request new weather data from the service
     private boolean updateWeatherData() {
+        if (bridge == null) {
+            logger.warn("Unable to update weather data. Bridge missing.");
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -228,10 +267,58 @@ private void updateChannel(String channelId) {
+
+        MeteoBlueBridgeHandler handler = (MeteoBlueBridgeHandler) bridge.getHandler();
+        if (handler == null) {
+            logger.warn("Unable to update weather data. Handler missing.");
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/internal/MeteoBlueConfiguration.java:
```diff
@@ -30,20 +32,20 @@
+    // longitude - the longitude of this location in degrees (-180 to 180)
+    public Double longitude;
     // altitude - the height above sea level of the location, in meters
     public Integer altitude;
```
Nice catch. Fixed.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueBridgeHandler.java:
```diff
@@ -0,0 +1,79 @@
+        MeteoBlueBridgeConfig config = getConfigAs(MeteoBlueBridgeConfig.class);
+        String apiKeyTemp = config.getApiKey();
+        if (StringUtils.isBlank(apiKeyTemp)) {
+            logger.warn("Cannot initialize Meteoblue bridge. No apiKey provided.");
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -93,31 +98,63 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+     * Marks the configuration as invalid.
+     */
+    private void flagBadConfig(String message) {
+        logger.warn(message);
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -110,12 +104,12 @@
 			<channel id="snowFall" typeId="snowFall" />
 			<channel id="cardinalWindDirection" typeId="cardinalWindDirection" />
 			<channel id="windDirection" typeId="windDirection" />
```
So, for this, I would need to remove the entire `<channel-type>` block for each affected item?

Delete this:

	<channel-type id="pressureMax">
		<item-type>Number:Pressure</item-type>
		<label>High Pressure</label>
		<description>High Pressure</description>
		<category>Pressure</category>
		<state readOnly="true" pattern="%.2f %unit%" />
	</channel-type>

And change this:

	<channel id="maxSeaLevelPressure" typeId="pressureMax" />

To this?

	<channel id="maxSeaLevelPressure" typeId="barometric-pressure" />

triller-telekom replied:
Almost, you have to remove the channel-type definition and change your reference to

```
<channel id="maxSeaLevelPressure" typeId="system.barometric-pressure" />
```

9037568 replied:
After doing this, I get warnings for the system.outdoor-temperature type:

````
07:58:48.744 [WARN ] [core.internal.thing.ThingTypeResource] - Cannot find channel type: system:outdoor-temperature
````

9037568 replied:
And in addition, it appears none of these types work.

![image](https://user-images.githubusercontent.com/9916473/43045509-d288a00a-8d87-11e8-984b-dd0099a83a92.png)

triller-telekom replied:
Do you have the recent commit https://github.com/eclipse/smarthome/commit/7bb5ac0474a2736ee92aabda6307fd4e390b4bd9 from the master branch in your branch? If not, please rebase yours onto the current master and try again.

9037568 replied:
I just realized why it's not working. I'm running the 2.3.0 OH2 release, which doesn't have this commit.
So I can either add these type changes without testing them, or leave them out.

triller-telekom replied:
Where are you developing your binding? Inside the ESH IDE? Or an openHAB IDE? Since you created a PR against ESH directly, I am assuming the former. In there you can easily rebase your branch and start ESH with your bundle, right?

If you are inside the openHAB IDE you can edit the run configuration, disable the `org.eclipse.smarthome.core.thing` bundle from the target platform and select it from the checked out project in your work space which should then of course be based on the current master from ESH.

Does that help?

9037568 replied:
If you're referring to [this](https://docs.openhab.org/developers/development/ide.html), I've never been able to get a functional workspace from it. Thousands of unresolvable bogus errors.

So I've never been able to use an IDE for debugging...

triller-telekom replied:
In your case it would even be sufficient to use the ESH IDE: https://www.eclipse.org/smarthome/documentation/development/ide.html

I have followed these steps just last week to install Photon and it worked with just one small issue: After following all steps I had to manually clean the projects with errors and let eclipse rebuild them twice or maybe even 3 times. In case Paper UI doesnt work out of the box, just run "PaperUI Setup.launch" from the paperUi project again.

9037568 replied:
I'll give this a shot and see what I get...

9037568 replied:
So not good.

After running all of the steps, I reach the one that says "Your workspace should now fully compile", and as usual, it does not:

![image](https://user-images.githubusercontent.com/9916473/43199118-eed7c598-8fde-11e8-82f2-1371a4b1457b.png)

![image](https://user-images.githubusercontent.com/9916473/43199146-0961fbf4-8fdf-11e8-815f-c80662791948.png)

triller-telekom replied:
Please see https://github.com/eclipse/smarthome/pull/4821#discussion_r205007871

I had to clean the projects with errors several times too, until they finally compiled.

9037568 replied:
How many dozens of cleans are we going to say is appropriate?
I've done several and the number of errors never declines below 286.

triller-telekom replied:
Strange... For me it worked. What you can try is to open the `org.eclipse.smarthome.model.codegen` project and let the `Generate All Models.launch` script run again (ignore that projects might have errors). And afterwards clean the projects with errors one more time.

If this works, we probably should update our documentation, that after cleaning and rebuild these issues can be solved by running the model generation again.

9037568 replied:
If I right-click any single project and try to build it, I get this:

![image](https://user-images.githubusercontent.com/9916473/43456340-2608dec4-9491-11e8-92a0-23e8546d2f42.png)

After I closed all the projects and opend the `codegen` project, it built successfully. I then ran the `Generate All Models.launch` script. It generated a number of errors like this:

````
Could not load class: org.eclipse.core.runtime.OperationCanceledException
Add org.eclipse.equinox.common to the class path.
````

I then reopened all the projects. After rebuilding, the number of errors is now 379.
Wrong direction.

triller-telekom replied:
I am sorry to say that I am out of ideas for now...

Any chance of using the latest OH 2.4.0-SNAPSHOT for testing, since you said you have 2.3.0?

Otherwise I would say that you leave out these new system channels and someone can come up with a follow-up PR.

9037568 replied:
At this point, I'm going to advocate for leaving them out.
If someone thinks they're useful, they can add them later.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -41,41 +45,31 @@
+				<context>location</context>
 				<label>Location</label>
-				<description>Name of this location</description>
+				<default>51.48,0.00,46</default>
```
Can you please comment on my remark here?

triller-telekom replied:
If you do not like to implemented the suggested use of the system location, OK. But I also do not like the idea to have a default which points to a location that is not of interest for the user at all. So I would vote for removing the default.

9037568 replied:
Removed it.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -41,41 +45,31 @@
+				<description>A symbolic, arbitrary name for the location</description>
 			</parameter>
-			<parameter name="timeZone" type="text" required="false">
+			<parameter name="timeZone" type="text">
```
I don't know why github marks this as outdated, but my comment is still valid. Can you please have a look?

9037568 replied:
Done.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -442,48 +318,26 @@
 		<label>Rain Spot</label>
-		<description>Precipitation distribution around the location. 7x7 array encoded
-		from south to north and west to east.</description>
+		<description>Precipitation distribution around the location as a 7x7 array,
```
see above, still valid

9037568 replied:
Done.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -228,10 +267,58 @@ private void updateChannel(String channelId) {
         }
     }
 
+    private State getStateForType(String type, Integer value) {
```
Still valid.

9037568 replied:
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue.test/.classpath:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I had intended to write some tests. If you can recommend a set to use as reference I can add some this weekend.

kaikreuzer replied:
What kind of tests make sense usually depends a lot on the binding. You might want to look at the [hue tests](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/java/org/eclipse/smarthome/binding/hue). If you don't find anything appropriate to test, this is also acceptable, but clearly tests are very welcome!

9037568 replied:
Working on this...

kaikreuzer replied:
Any update here? The test fragment still looks empty.

9037568 commented:
Thanks for the review, @kaikreuzer. I think I've understood them all, and will try to knock these out this weekend.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue.test/pom.xml:
```diff
@@ -0,0 +1,125 @@
+  <version>0.10.0-SNAPSHOT</version>
+  <packaging>eclipse-test-plugin</packaging>
+
+  <name>MeteoBlue Binding Tests</name>
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,18 @@
+		xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+		<!-- Meteoblue Bridge -->
+		<bridge-type id="bridge">
+				<label>Meteoblue Bridge</label>
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,18 @@
+		<!-- Meteoblue Bridge -->
+		<bridge-type id="bridge">
+				<label>Meteoblue Bridge</label>
+				<description>The meteoblue bridge holds settings for use with the Meteoblue weather service.</description>
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,20 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name:  MeteoBlue Binding Tests
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+				xmlns:binding="http://eclipse.org/smarthome/schemas/binding/v1.0.0"
+				xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+	<name>MeteoBlue Binding</name>
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,161 @@
+
+## Full Example
+
+_Provide a full usage example based on textual configuration files (*.things, *.items, *.sitemap)._
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/pom.xml:
```diff
@@ -0,0 +1,19 @@
+	<artifactId>org.eclipse.smarthome.binding.meteoblue</artifactId>
+	<version>0.10.0-SNAPSHOT</version>
+
+	<name>MeteoBlue Binding</name>
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueBridgeHandler.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Override
+    /**
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueBridgeHandler.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Override
+    /**
```
Done.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,161 @@
+| ------------- |:-------------:| :-------: | -------------------- |
+| serviceType   |               | Yes       | The service type to be used.  Either 'Commercial' or 'NonCommercial' |
+| location      |               | Yes       | The latitude, longitude, and optionally altitude of the location, separated by commas (e.g. 45.6,45.7,45.8). Altitude, if given, should be in meters.
+| locationName  |               | No        | A symbolic, arbitrary name for the location |
```
That's what I mean: If it passes back exactly what you have passed in, there isn't any point in having this channel.

9037568 replied:
Ok. Removed.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,421 @@
+		</channel-groups>
+
+		<config-description>
+			<parameter name="serviceType" type="text" required="true">
```
Done. Corrected in README as well.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+        }
+
+        if (StringUtils.isBlank(config.location)) {
+            flagBadConfig("The location setting was not configured.");
```
Done.

wborn commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue.test/pom.xml:
```diff
@@ -0,0 +1,125 @@
+  <properties>
+    <bundle.symbolicName>org.eclipse.smarthome.binding.meteoblue.test</bundle.symbolicName>
+    <bundle.namespace>org.eclipse.smarthome.binding.meteoblue</bundle.namespace>
+  </properties>
```
Please remove these properties. See #6060.

9037568 replied:
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,33 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ClassPath: .
+Import-Package: 
+ javax.annotation;resolution:=optional,
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+        MeteoBlueConfiguration config = getConfigAs(MeteoBlueConfiguration.class);
+
+        if (StringUtils.isBlank(config.serviceType)) {
+            flagBadConfig("The serviceType setting was not configured.");
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+            return;
+        }
+
+        Runnable runnable = new Runnable() {
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,161 @@
+
+| Group Name       |
+| ---------------- |
+| forecastToday    |
```
Ick. Ick. Ick.
But done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,161 @@
+    Text item=weatherLoc
+    Text item=weatherLat
+    Text item=weatherLon
+    Group item=weatherDay0
```
Fixed.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+                            updateChannel(channel.getUID().getId());
+                        }
+                    }
+                } catch (Exception e) {
```
Changed to WARN.

9037568 commented:
I still have a few review items left to address...

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,161 @@
+| precipitationHours       | Number             | Total hours of the day with precipitation |
+| humidityGreater90Hours   | Number             | Total hours of the day with relative humidity greater than 90% |
+
+The following channels are not specific to any day's forecast.
```
So `locationName` has already been removed.
`latitude`, `longitude`, `height` and the others, I'm going to argue for keeping because they could be useful to have.

triller-telekom replied:
Regarding these 3 channels: latitude, longitude, height. **All** of them are included in location, so whats the point for introducing redundancy here?

9037568 replied:
`location` is only used in configuration.
If someone wants to display the latitude and longitude, they have to exist as channels, right? If there's a better way, fine, but I'm not seeing it...

triller-telekom replied:
As kai already mentioned. You put the information about lat/lon into the thing configuration and this input would be the output of you channels. This doesn't make much sense, i.e. it doesn't create any value. If a user wants to display the coordinates somewhere he can easily take those he entered into the configuration.

Channels in a Thing are supposed to describe functionalities of a Thing. Printing out a configuration value is not a functionality of the Thing.

kaikreuzer replied:
This still seems to be open.

9037568 replied:
I've removed the following channels: latitude, longitude, timezoneAbbreviation, utcTimeOffset, modelRunUTC, and modelRunUpdateTimeUTC.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "API Key not authorized for this datafeed");
+            } else {
+                logger.warn("Failed to retrieve weather data due to unexpected error. Error message: {}", errorMessage);
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+        // Build a State from this value
+        State state = null;
+        if (datapoint == null) {
+            state = UnDefType.UNDEF;
```
Updated to not send a state update if the datapoint is null.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+            }
+            return false;
+        }
+
```
Done.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+            logger.trace("http response: {}", httpResponse);
+            return httpResponse;
+        } catch (IOException e) {
+            logger.warn("I/O Exception occurred while retrieving weather data.", e);
```
Changed to DEBUG and set status to OFFLINE.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/internal/Forecast.java:
```diff
@@ -0,0 +1,565 @@
+        }
+
+        pictocode = dataDay.getPictocode()[whichDay];
+        icon = getIconNameForPictocode(pictocode);
```
I had a reason for this at some point. Can't remember it now. It's currently delivering the filename for the image file.
The actual icon is displayed in the sitemap using this construction using the pictocode channel:

````
Number todayPCode    "Pictocode [%d]"  <iday>  (weatherDay0)  {channel="meteoblue:weather:LXK:forecastToday#pictocode"}
````

So I'm not sure this icon channel is needed...

kaikreuzer replied:
From a user functional perspective the image channel makes much more sense than a number channel, so this should be fixed and the number channel removed.

9037568 commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueBridgeHandler.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        apiKey = apiKeyTemp;
+        updateStatus(ThingStatus.ONLINE);
```
Removed.

lolodomo commented:
The merge of this new binding would be appreciated as the WU binding is apparently no more working.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/build.properties:
```diff
@@ -0,0 +1,8 @@
+             OSGI-INF/,\
+             ESH-INF/,\
+             NOTICE
+additional.bundles = org.eclipse.pde.ds.lib
```
What is this about? Have never seen this so far...

9037568 replied:
Not a clue. Removed.

9037568 replied:
And after reviewing in Eclipse, I remember. Eclipse is issuing this warning after this line is removed:

````
DS Annotations missing from permanent build path
````

kaikreuzer replied:
None of the other bundles have this. So I guess you are not using the official IDE setup.
Please make sure that you have the "add DS Annotations to classpath" checked:

![screen shot 2018-09-26 at 11 25 58](https://user-images.githubusercontent.com/3244965/46070851-0beeb200-c17f-11e8-9411-1a9728527ad3.png)

9037568 replied:
Actually it is the official setup. But it was set up 2-3 years ago.
I'll look into the preferences later. Thanks for the pointer.

9037568 replied:
Just for the record, this setting was already enabled for both the project and the workspace. Doesn't help at all.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+            return;
+        }
+
+        updateStatus(ThingStatus.ONLINE);
```
Still open.

9037568 replied:
Done.

kaikreuzer commented:
Thanks for the updates, so mainly https://github.com/eclipse/smarthome/pull/4821#discussion_r209188628 is still left.

9037568 commented:
@kaikreuzer I've made the last changes for the Image, but I found while attempting to test them that the binding is now broken as the bridge never initializes.

It appears this is caused by [this change](https://github.com/eclipse/smarthome/pull/4821/commits/853032237a518bb4988a9db2c50b3e28e3f869d6). Can you advise how to fix without restoring the line you asked to have removed?

kaikreuzer commented:
I assume you refer to [this comment](https://github.com/eclipse/smarthome/pull/4821#discussion_r209180177). Please read it completely, I also stated "It should be this handler, which is responsible of keeping the connection to the service.".

This does not seem to be the case, I cannot see the handler trying to establish a connection to the web service. Only if this is successful, the bridge should be set to ONLINE. If the network is not available or the api key is invalid, it should instead go to OFFLINE.

9037568 commented:
Ok, @kaikreuzer, I think I'm almost there. I updated the bridgehandler to do a healthcheck before setting its status to ONLINE. I also fixed the icon channel to return an Image instead of a string.

However, I think the pictocode channel with the numeric value is still needed, as I can't find a way to get this displayed in BasicUI without it:

![image](https://user-images.githubusercontent.com/9916473/46248982-62622780-c3ef-11e8-8234-7d9319f9176e.png)

Here's the snippet from the items file:

````
Number todayPCode    "Pictocode [%d]"  <iday>  (weatherDay0)  {channel="meteoblue:weather:metBridge:LXK:forecastToday#pictocode"}
````

kaikreuzer commented:
I think your example nicely shows that it does not make sense that way. You now see "2" as a state for this item, which is useless for the user.
What would make sense is to rename the channel to something like "weatherCondition" and have state options provided with it, which would provide suitable label texts, e.g. like "sunny", "partly cloudy" etc. Then this channel would not be anymore about a "pictogram id", but about providing an overall weather condition. Does such a mapping exist?

9037568 commented:
Sounds like an enumeration of states? I think I see how to do that...

9037568 commented:
Success! I have a bit of code cleanup to do, but the new condition channel appears to be working.

![image](https://user-images.githubusercontent.com/9916473/46517759-c137fe80-c83e-11e8-972a-6c313326ee3e.png)

kaikreuzer commented:
Great, that looks nice 👍

9037568 commented:
That last commit should complete this.

9037568 commented:
@kaikreuzer just tagging you to make sure this gets flagged as ready for your review.

kaikreuzer commented:
Thanks, I saw it. I'm currently on holiday and will look at your updates next week when I am back, promised!

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,388 @@
+	</channel-type>
+
+	<channel-type id="condition">
+		<item-type>String</item-type>
```
Actually, your value seems to be a Number, wouldn't that be the better match?

kaikreuzer replied:
ok

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/README.md:
```diff
@@ -0,0 +1,151 @@
+| minSeaLevelPressure      | Number:Pressure    | Low sea level pressure  |
+| maxSeaLevelPressure      | Number:Pressure    | High sea level pressure |
+| meanSeaLevelPressure     | Number:Pressure    | Mean sea level pressure |
+| condition                | String             | A brief description of the forecast weather condition (e.g. 'Overcast') |
```
Maybe worth to mention that the state is a number between 1-17 (and provide the link to https://content.meteoblue.com/nl/service-specifications/standards/symbols-and-pictograms#eztoc14635_1_6 here).

9037568 replied:
Done.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/src/main/java/org/eclipse/smarthome/binding/meteoblue/handler/MeteoBlueHandler.java:
```diff
@@ -0,0 +1,425 @@
+    @Override
+    public void dispose() {
+        logger.debug("Disposing MeteoBlue handler.");
+        shuttingDown = true;
```
Any comment on that one?

9037568 replied:
Ok, I removed it.

9037568 commented:
@kaikreuzer I think this is back to you again...

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,421 @@
+		<label>Rain Spot</label>
+		<description>Precipitation distribution around the location as a 7x7 array,
+		encoded from south to north and west to east
+		(e.g. 2112211093222219212110112229922133311912239022192).</description>
```
Because this is raw technical data, which does not make any sense for a user to e.g. display in a UI. For this, we have the Image channel, which makes sense. We must not add channels just because of the sake of having some random data at hand.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.meteoblue/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,33 @@
+Bundle-ClassPath: .
+Import-Package: 
+ javax.measure.quantity,
+ com.google.common.collect,
```
We are not recommending the use of Guava anymore - please remove this dependency and use Java8 constructs instead.

kaikreuzer commented:
@9037568 Afaics, only https://github.com/eclipse/smarthome/pull/4821#discussion_r229999400 is left open to discuss, correct?

9037568 commented:
Yes, that's all. The data in question is useful to me, and I don't want it removed.
I'm planing to use it data for local 3d renders. I just haven't had time to do any so far...

kaikreuzer commented:
Ok @9037568, then let's declare this ready for a merge.
We will need to create a CQ to get approval by the IP team and for this you will need to squash your commits into a single one, please.

9037568 commented:
Thanks! I will do the squash tonight.

9037568 commented:
Ok, @kaikreuzer, we're all squashed here.

9037568 commented:
I fixed the conflict, but github still claims there's a conflict. What do I have to do, rebase to master?

9037568 commented:
And because 3 squashes just aren't enough, apparently we have to do it again...

kaikreuzer commented:
@9037568 Please never ever do merge commits! Always do a proper "rebase on master" instead.
I've now cleaned up your branch and did a force push to have a single commit with no conflicts and no merge commit. Please refrain from doing any further changes now as I'll file a CQ next.

9037568 commented:
>Please never ever do merge commits!

Thanks, Kai. News to me. Hopefully I can remember it.

kaikreuzer commented:
Good news, we have check-in approval!

9037568 commented:
Awesome! Thanks!

9037568 commented:
I'd say deleting it is best. I ran out of time for writing new code a while back...
